File: Python3/Hard/WrongAnswer/kth-smallest-product-of-two-sorted-arrays_2.py
class Solution:
    def kthSmallestProduct(self, nums1: List[int], nums2: List[int], k: int) -> int:
        
        # binary search
        def count_smaller(x):
            # count the number of elements in nums1 * nums2 smaller than x
            # two pointer
            j = len(nums2) - 1
            cnt = 0
            for i in range(len(nums1)):
                while j >= 0 and nums1[i] * nums2[j] >= x:
                    j -= 1
                cnt += j + 1
            return cnt
        
        # find the smallest x s.t. count_smaller(x) >= k
        left, right = 0, 10**10
        while left < right:
            mid = (left + right) // 2
            if count_smaller(mid) >= k:
                right = mid
            else:
                left = mid + 1
        return left

File: C++/Hard/WrongAnswer/kth-smallest-product-of-two-sorted-arrays_2.cpp
class Solution {
public:
    long long kthSmallestProduct(vector<int>& nums1, vector<int>& nums2, long long k) {
        
        long long n = nums1.size(), m = nums2.size();
        if (n > m) return kthSmallestProduct(nums2, nums1, k);
        long long l = nums1[0] * nums2[0], r = nums1[n - 1] * nums2[m - 1];
        while (l < r) {
            long long mid = l + (r - l) / 2;
            long long cnt = 0;
            for (int i = 0, j = m - 1; i < n; ++i) {
                while (j >= 0 && (long long)nums1[i] * nums2[j] > mid) --j;
                cnt += j + 1;
            }
            if (cnt < k) l = mid + 1;
            else r = mid;
        }
        return l;
    }
};